test/UpgradeRegressionManager.invariant.t.sol

```diff
@@ -0,0 +1,122 @@
+  SimpleContract public target;
+  UpgradeRegressionManagerHandler public handler;
+
+  address public admin = makeAddr("admin");
+  address public guardian = makeAddr("guardian");
+
+  function setUp() public {
+    timelockTarget = new MockTimelockTarget();
+    timelockTarget.setDelay(5 seconds);
+
+    target = new SimpleContract();
+
+    urm = new UpgradeRegressionManager(ITimelockTarget(timelockTarget), admin, guardian, 5 seconds);
+
+    handler = new UpgradeRegressionManagerHandler(urm, admin, guardian, target);
+
+    // target the handler for invariant testing
+    targetContract(address(handler));
+  }
+
+  /*///////////////////////////////////////////////////////////////
+                        Rollback State Invariant
+  //////////////////////////////////////////////////////////////*/
+
+  /// @notice A rollback cannot exist in both `rollbackQueueExpiresAt` and `rollbackExecutableAt` simultaneously
+  /// @dev Rationale: Ensures rollback is in exactly one state at a time
+  function invariant_rollbackStateMutualExclusion() public view {
+    uint256 rollbackId = handler.getRollbackId();
+
+    bool inQueueingState = urm.rollbackQueueExpiresAt(rollbackId) != 0;
+    bool inExecutionState = urm.rollbackExecutableAt(rollbackId) != 0;
+
+    assertFalse(inQueueingState && inExecutionState, "Rollback cannot exist in both queueing and execution state");
+  }
+
+  /*///////////////////////////////////////////////////////////////
+                        Timing Invariant
+  //////////////////////////////////////////////////////////////*/
+
+  /// @notice `rollbackQueueWindow` must always be greater than 0
+  /// @dev Ensures valid queue window for rollbacks
+  function invariant_rollbackQueueWindowPositive() public view {
+    uint256 _rollbackQueueWindow = urm.rollbackQueueWindow();
+    assertGt(_rollbackQueueWindow, 0, "Rollback queue window must be greater than 0");
+  }
+
+  /// @notice When a rollback is queued, its ETA must be in the future
+  /// @dev Ensures proper timing for execution
+  function invariant_etaInFutureAfterQueue() public view {
+    uint256 _rollbackId = handler.getRollbackId();
+    uint256 _eta = urm.rollbackExecutableAt(_rollbackId);
+    uint256 _currentTime = block.timestamp;
+
+    // nothing to test — either not proposed or still within queue window
+    if (_eta == 0) {
+      return;
+    }
+
+    assertGt(_eta, _currentTime, "Rollback ETA must be in the future");
+  }
+
+  /// @notice Cannot queue a rollback after its `rollbackQueueExpiresAt` has passed
+  /// @dev Enforces queue window constraints
+  function invariant_queueFailsAfterExpiry() public {
+    uint256 _rollbackId = handler.getRollbackId();
+    uint256 _expiresAt = urm.rollbackQueueExpiresAt(_rollbackId);
+
+    // nothing to test — either not proposed or still within queue window
+    if (_expiresAt == 0 || block.timestamp < _expiresAt) {
+      return;
+    }
+
+    try handler.queueRollback() {
```

Comment:
The fact we have to do this is a smell that says this doesn't make sense as an invariant. A proper invariant test should not try to take an action like this. It should just assert some reality about the states of the system.